Sources/ASTPreprocessor/ASTPreprocessor.swift

```diff
@@ -0,0 +1,67 @@
+//
+//  ASTPreprocessor.swift
+//  ASTPreprocessor
+//
+//  Created by Nik Vangerow on 11/22/18.
+//
+
+import AST
+
+/// The Preprocessing pass for the AST.
+public struct ASTPreprocessor: ASTPass {
+
+  public init() {}
+
+  // Make binary expressions involving the dot (.) operator left-associative.
+  // Binary expressions are parsed with the wrong associativity. Due to the recursive descent parsing,
+  // Expressions associate to the right: a.(b.c). This is wrong. We want them to associate to the right: (a.b).c.
+  public func process(binaryExpression: BinaryExpression,
+                      passContext: ASTPassContext) -> ASTPassResult<BinaryExpression> {
+    guard binaryExpression.opToken == .dot else {
```

Comment:
Perhaps have a `switch` to match on the case of the `.` operator instead of `guard` + `return`? Then also put the comment for the function to that specific case. Also there could be other operations we want to be right-associative, e.g. `**` (exponentiation), so the `Token` could have a function for associativity (currently it has a similar function for precedence).

Reply:
Why Switch? We are not handling any other case so guard is a lot cleaner. It's certainly swiftier. If someone wants to handle another case (though why would they?) they can always refactor it then.

Reply:
in addition by your suggestion we'd nest the second guard within the switch. Ew.

Reply:
Fair enough, but I'd still rather make it easy to mark any binary operation right-associative.

Reply:
We should do this on an on-demand basis. Exponentiation (if it exists) should work fine

Reply:
Exponentiation does exist and it should be right-associative (it is the canonical example of a RA binary operation after all), but I guess there are no tests for it.

Reply:
If we want to fix exponentiation then that should be a separate ticket. This ticket is for .

Reply:
Yes, don't fix the associativity for it, but no reason to make the `ASTPreprocessor` framework completely `.`-specific.

Reply:
Ok I'll open a separate ticket for this for the backlog.